Repair some Gitea accessibility issues

fsologureng opened the pull request:
**Updated 2023-03-22 removing old ideas and merges done apart.**

This PR is an improvement of #21743 taking the combobox+listbox ARIA design pattern as a better fit for some Fomantic dropdowns, with the addition of other a11y improvements.

The idea is to resolve the *main problem* described in #21742.

Changed foomantic dropdowns to the combobox+listbox ARIA design pattern
 - Changed scope application of aria.js functions to apply it after all widgets are initialized.
 - Fixes and improve a11y structure of:
   - Reactions in issue/pr comments view.
   - Comment options in issue/pr comments view.
   - Milestone dropdown in repo issue/pr view.
   - Projects dropdown in repo issue/pr view.
   - Assignees dropdown in repo issue/pr view.
   - Reviewers dropdown in repo pr view.
   - Due date dropdown in repo issue/pr view.
   - Dependency dropdown in repo issue/pr view.
   - Reference copy button in repo issue/pr view.
   - Language selector.
   - Create... button.
   - Profile and settings... button.
 - Fixes lack of aria label of Close issue button.
 - Add label for some buttons.
 - Remove `dropdown` class to an svg which generates false positives

silverwind commented on templates/repo/settings/options.tmpl:
```diff
@@ -23,20 +23,20 @@
-						<input name="template" type="checkbox" {{if .Repository.IsTemplate}}checked{{end}}>
-						<label>{{.locale.Tr "repo.template_helper"}}</label>
+						<input id="template" name="template" type="checkbox" {{if .Repository.IsTemplate}}checked{{end}}>
+						<label for="template">{{.locale.Tr "repo.template_helper"}}</label>
```
Can we nest the `<input>` inside the label in these cases without breaking the rendering? If yes, it would be preferred as that method does not introduce new ids while being equally accessible.

fsologureng replied:
Hi.
The rendering remains (almost) the same, but the functionality is totally broken because the initialisation [can't find the `<input>`](https://github.com/go-gitea/gitea/blob/main/web_src/fomantic/build/semantic.js#L1243). Sorry, it was the first approach I tried.

[Linked labels](https://fomantic-ui.com/modules/checkbox.html#linked-labels) is the only option to make foomantic checkbox accessible.

silverwind replied:
I see, `.children` won't work recursively, so it's not a solution that can work.

silverwind commented:
This looks like a bunch of good changes, sorry for not picking it up earlier.

Maybe @wxiaoguang can have a look too.

wxiaoguang commented:
> This looks like a bunch of good changes, sorry for not picking it up earlier.
> 
> Maybe @wxiaoguang can have a look too.

I do not like the idea of writing a lot of unnecessary IDs in the templates. Quote my original thoughts here.

> From a development view, I do not think it's easy to make every developer(contributor) understand a11y clearly and write all dropdown elements with correct aria-attributes, and Fomantic UI Dropdown is quite special. In the end, attachOneDropdownAria is still widely used and it is a general solution, then the templated dropdowns will become a one-time code just for the reactions. Then why not just keep one general attachOneDropdownAria without manually written IDs?

Although generating IDs dynamically affects performance slightly, I think it's acceptable and better than writing unnecessary IDs everywhere.

And all Fomantic checkbox/radio inputs use the `<label /><input />` structure now, I do not feel good about adding IDs for all of them, it will  be a lot , and make it difficult to refactor them (replace them) in the future.

In the future, these components could be replaced by  some well-designed a11y  ones, still do not force  developers to manually write every ID for the elements in dropdowns.


Just my opinion, not a blocker (if the code works)

silverwind commented:
Yeah I already commented above on the unnecessary IDs. If we can get the rendering to work with nesting, it's definitely the way to go. If this change is too massive, I would suggest factoring it out into a separate PR to ease review here.

fsologureng commented:
> Although generating IDs dynamically affects performance slightly, I think it's acceptable and better than writing unnecessary IDs everywhere.

I don't understand why you say _unnecessary IDs everywhere_, are all those IDs added by aria.js unnecessary? If they are unnecessary, why are they added?

If the current markup of this input+labels pair are not accessible, why would it be more difficult to change them with id+for attributes added towards a nested schema? sounds illogical; that's a deep change with or without attributes, includes templates, css and UI framework.

> In the future, these components could be replaced by some well-designed a11y ones, still do not force developers to manually write every ID for the elements in dropdowns.

_In the future_ for a11y solutions is practically never. Why not implement accessible solutions as soon as possible?
Why are developers and contributors exempt from knowing about accessibility? Why not about css, javascript or UI frameworks? Why shouldn't a **web developer** know about ARIA?
ARIA was invented to customise web controls and user interaction, to refresh the static and boring accessible world of simple selections and ugly checkboxes. Why not embrace it? It's pure web design. It's not that difficult.

In many institutions it is essential to provide accessibility, even in financial institutions, so why is it seemingly unnecessary here?

wxiaoguang commented:
> > Although generating IDs dynamically affects performance slightly, I think it's acceptable and better than writing unnecessary IDs everywhere.
> 
> I don't understand why you say _unnecessary IDs everywhere_, are all those IDs added by aria.js unnecessary? If they are unnecessary, why are they added?
> 

The manually-written IDs are unnecessary, it can be handled by a general approach by a framework. Fomantic UI just does it wrongly, it's not ideal introduce a lot of IDs to fix the Fomantic UI's fault, in the end there will be a mess.

> If the current markup of this input+labels pair are not accessible, why would it be more difficult to change them with id+for attributes added towards a nested schema? sounds illogical; that's a deep change with or without attributes, includes templates, css and UI framework.

This situation could be handled by a general approach automatically, instead of writing them again and again manually.

> > In the future, these components could be replaced by some well-designed a11y ones, still do not force developers to manually write every ID for the elements in dropdowns.
> 
> _In the future_ for a11y solutions is practically never. Why not implement accessible solutions as soon as possible?

Is your solution a 100% complete solution? Do you promise to pay all attention to all future commits to make sure all of them work properly with a11y? If no, why not try to use a general approach, eg: adding IDs automatically?

> Why are developers and contributors exempt from knowing about accessibility? Why not about css, javascript or UI frameworks? Why shouldn't a **web developer** know about ARIA? ARIA was invented to customise web controls and user interaction, to refresh the static and boring accessible world of simple selections and ugly checkboxes. Why not embrace it? It's pure web design. It's not that difficult.

If yes, why the code looks like this at the moment? You can try to review the recent frontend PRs to see whethe most developers and contributors really knows a11y well or care about it.

> In many institutions it is essential to provide accessibility, even in financial institutions, so why is it seemingly unnecessary here?

Again, IMO a general approach is better, I do not think manually writing all IDs is a proper approach. And again, just my opinion.

wxiaoguang commented:
More update: why this problem is more serious in Gitea? Because other projects are more modern, their UIs just work well and developers do not need to pay much attention to a11y. BUT, the birth of Gitea is too early, and Gitea is still using Fomantic UI now, the Fomantic UI is badly designed for a11y, so all problems come.

To resolve the problem fundamentally, IMO there could be an intermediate framework to fix a11y problems automatically, and then replace buggy Fomantic UI components with modern/native ones -- instead of forcing developers to write IDs/aria-labels again and again.

fsologureng commented:
> Do you promise to pay all attention to all future commits to make sure all of them work properly with a11y?

I will do my best effort to provide accessibility.

> why the code looks like this at the moment?
Because there are strong incentives in order to develop without taking in account accessibility.

------
> To resolve the problem fundamentally, IMO there could be an intermediate framework to fix a11y problems automatically, and then replace buggy Fomantic UI components with modern/native ones -- instead of forcing developers to write IDs/aria-labels again and again.

So developers will never have to deal with IDs nor ARIA stuff in templates, but they have to develop that intermediate framework you propose, dealing with IDs and ARIA stuff on pure javascript. The problem I see is who is going to develop that intermediate framework if, in your words, developers aren't called to deal with ARIA stuff?

Also, developers will have to choose a good UI framework that provide good accessibility and implement that framework on top of gitea, but without knowing too much about a11y, it sounds a complex decision. Who, without knowing too much about a11y, are willing to choose a good UI framework like the one you describe for the whole frontend of gitea while ensuring accessibility?

On the other hand, with your last statements, I start to understand that aria.js is a kind of prototype of that intermediate framework that frees developers from dealing with ARIA and IDs in templates and fixes foomantic UI bugs in dropdowns. But I have opened a [bug](#21742) a month ago and nobody (zero) person has even reacted (less confirmed) anything about it. What is the commitment to this approach?

Finally, the checkboxes on the configuration page don't have an accessibility problem due to foomantic UI, it's a simple case of malformed html (not the only one in that template). If you want to preserve a malformed html template while you find a good framework fixing the html with javascript, well I'm not going to provide it. It's easy to see why checkboxes are siblings with their labels in foomantic (and in other frameworks like Vue, too). Adding the ID and FOR attrs is a straightforward native solution if you remove the fomantic UI from them.

wxiaoguang commented:
> So developers will never have to deal with IDs nor ARIA stuff in templates, but they have to develop that intermediate framework you propose, dealing with IDs and ARIA stuff on pure javascript. The problem I see is who is going to develop that intermediate framework if, in your words, developers aren't called to deal with ARIA stuff?

Nope. IMO the intermediate framework should/could/only handle most bad cases, for example, then most developers do not need to manually write IDs again and again. Everything should just works like modern HTML frameworks.

> Also, developers will have to choose a good UI framework that provide good accessibility and implement that framework on top of gitea, but without knowing too much about a11y, it sounds a complex decision. Who, without knowing too much about a11y, are willing to choose a good UI framework like the one you describe for the whole frontend of gitea while ensuring accessibility?
> On the other hand, with your last statements, I start to understand that aria.js is a kind of prototype of that intermediate framework that frees developers from dealing with ARIA and IDs in templates and fixes foomantic UI bugs in dropdowns. But I have opened a [bug](#21742) a month ago and nobody (zero) person has even reacted (less confirmed) anything about it. What is the commitment to this approach?

I have complained similar things before when I was doing refactoring works for Gitea (you can take a look at my complete PRs about refactoring), then some people told me "that's how open-source project works", so now I won't touch the code unless I get enough supports first. 😂 . Just to share my personal feeling and opinion.

> Finally, the checkboxes on the configuration page don't have an accessibility problem due to foomantic UI, it's a simple case of malformed html (not the only one in that template). If you want to preserve a malformed html template while you find a good framework fixing the html with javascript, well I'm not going to provide it. It's easy to see why checkboxes are siblings with their labels in foomantic (and in other frameworks like Vue, too). Adding the ID and FOR attrs is a straightforward native solution if you remove the fomantic UI from them.

What's the difference between them? Do these "working" checkeboxes all have manually written IDs? Personally I won't say writing IDs for all checkboxes is a proper approach. The better one should be using `<lable><input></label>`, or adding IDs automatically. I think the only divergence between us is: you prefer to fix the problem by adding all IDs manually, while I prefer to fix the problem by some framework code and adding the IDs automatically. I also agree to improve the a11y problem ASAP, and if other people like, my opinion is never a blocker.

wxiaoguang commented:
The changes in this PR:

L-G-T-M part:

* The `role` / `aria-label` in head/footer: L-G-T-M
* The removal/change of `tabindex="..."`: L-G-T-M
* The `e.key === 'ESC'`: L-G-T-M

not L-G-T-M part:
* The manually added IDs: not L-G-T-M, especially the `IdPrefix`, I think it's too dirty.
* `data-aria-templated`, I haven't fully understood whether it's necessary. If not necessary (just some slightly performance difference), then it's not L-G-T-M.

Although I can not vote my approval for all these changes, at least, I can vote approval for the L-G-T-M part.

fsologureng commented:
@wxiaoguang what do you think about this aria.js refactory proposal?

wxiaoguang commented:
> @wxiaoguang what do you think about this aria.js refactory proposal?

Some problems have been addressed by other PRs, some problems have more details and backgrounds now, the comments in this PR are quite a  few, maybe we can start a fresh discussion based on the latest situation?

* I always agree to improve (anything)  if the refactoring doesn't introduce maintainability problems.
* For `aria.js`,  the major  problem of it at the moment is it always uses `menu + menuitem`, in the near future, if the dropdowns are still there, as a hacky patch, `aria.js` could use `combobox` instead of `menu` when the dropdowns have input elements.

---- 

About latest changes (just a quick and brief review)

* Could these `tabindex="0"` for buttons be removed? By default they are focus-able.
* About `js-role-combobox`
    * I think the "language" menu is a menu, it shows the menu item, if you click any one, then page refreshes, it doesn't look like a combobox which provides value for the form.
    * In most cases, IMO we do not need  the `js-role-combobox`, because if there is a `input.search` in dropdown, then it must be a combobox.
* `<div class="ui button" ...` could be `<button ...` then no need to add `tabindex`?
    * I think the `<button>` problem will be completely fixed by #23337 (@delvh)
* `<button class="delete-dependency-button tooltip ci muted"`
    * I am not sure about the style because I haven't tested this change, but I am sure `muted` is not needed because `muted`  is for `a` only.
* The `class="dropdown icon"`  classes on SVG elements in dropdowns  is done by purpose to fine-tune the UI alignment, there are CSS styles for them, I do not think the `dropdown` could be removed (I haven't tested, just a guess)
* `attachDropdownAria($('.ui.dropdown'));` doens't seem right. We should skip `custom`  dropdown elements, these `custom` elements have their own JS to work.
* The new JS code: I will take a look carefully later. A quick thought in my mind: maybe we need two functions, one for `menu` , one for `combobox` (they could share  some  code  by common functions), to make them independent, then any change on one side won't affect another one, and it's easier to maintain and test in the future.

wxiaoguang commented:
One more thing, the `item` shouldn't be clicked without any check. See the comment in: 
* #23343

fsologureng commented on templates/repo/issue/view_content/sidebar.tmpl:
```diff
@@ -661,7 +661,7 @@
 
 						<div class="text right actions">
-							<div class="ui cancel button">{{.locale.Tr "settings.cancel"}}</div>
+							<button class="ui cancel button">{{.locale.Tr "settings.cancel"}}</button>
```
I came here for asking why this change left the cancel button acting as submit button because I couldn't find an explanation debugging, but I saw #23337 and I am afraid I will need to revert it or wait to @delvh's pr being merged.

fsologureng commented:
> > @wxiaoguang what do you think about this aria.js refactory proposal?
> 
> Some problems have been addressed by other PRs, some problems have more details and backgrounds now, the comments in this PR are quite a few, maybe we can start a fresh discussion based on the latest situation?
> 
> (update: I hide my previous comments as out-dated, indeed they are out-dated)
> 
>     * I always agree to improve (anything)  if the refactoring doesn't introduce maintainability problems.
> 
>     * For `aria.js`,  the major  problem of it at the moment is it always uses `menu + menuitem`, in the near future, if the dropdowns are still there, as a hacky patch, `aria.js` could use `combobox` instead of `menu` when the dropdowns have input elements.
> 
> 
> About latest changes (just a quick and brief review)
> 
>     * Could these `tabindex="0"` for buttons be removed? By default they are focus-able.

done

> 
>     * About `js-role-combobox`
>       
>       * I think the "language" menu is a menu, it shows the menu item, if you click any one, then page refreshes, it doesn't look like a combobox which provides value for the form.

`menu` role doesn't have any element except the `menuitem` set. Language selector has the current value as a text which appears to be a readonly input. This is why I implemented as a combobox.

>       * In most cases, IMO we do not need  the `js-role-combobox`, because if there is a `input` in dropdown, then it must be a combobox.
>       * Update: according to new changes in `#23346`, I think deciding the role by `input` is doable.

Unfortunately, that approach is totally insufficient IMO. There are more structures than those two. In fact, the unique selectors which match a menu pattern are the reactions menu and the comment options menu (the only pair [you didn't fix with aria.js](https://github.com/go-gitea/gitea/issues/21742) at the beginning, btw). You can check this statement using the [Axe DevTools extension](https://www.deque.com/axe/browser-extensions/) with a fomantic dropdown opened. The majority of dropdowns have a kind of label in the middle incompatible with a menu pattern and have noise produced by elements as options headers and dividers which I am addressing here.

As I said in [Forgejo Dropdowns UX bug](https://codeberg.org/forgejo/forgejo/issues/383#issuecomment-811167) some days ago, keep talking about `dropdowns` as a whole IMO is a bad approach and is bad aligned with your proposal made in #23345. If you observe each widget implemented with Fomantic dropdown there are little differences that doesn't fit well in a rigid overview. That's why I didn't took the presence of an `input` as a final indicator of a `combobox`. The worst case is the tag/branch selector, which, as far as I can see, is a unique totally inaccessible structure.

This is the reason why I chosen to do specific identification of aria roles via `js-` prefixed classes and do a specific implementation accordingly. Is extensible and compatible with a step by step proposal. IMO the dependencies selector is the most near structure to an accessible combobox, I think that change some sidebar selectors to that structure will provide better accessibility of them.

The language selector is pretty close to be accessible, but auto-submit is a bad practice which can produce annoying interactions if you choose bad your desired language. Ideally you should be able to choose your language and then submit it to change the UI with clarity that the choosing was made without a mistake. That's the reason why combobox has a fillable field. If you want just a menu structure, you should put the current language label out of the selector or remove the label and only mark the current language in the corresponding menuitem (as a native selector does).

> 
>     * `<div class="ui button" ...` could be `<button ...` then no need to add `tabindex`?

done

>       
>       * I think the `<button>` problem will be completely fixed by [Convert `<div class="button">` to `<button class="button">` #23337](https://github.com/go-gitea/gitea/pull/23337) (@delvh)

I made a change here that needs to wait a merge of that pull request or be reverted because is failing currently in the cancel button of the locking issue modal.

> 
>     * `<button class="delete-dependency-button tooltip ci muted"`
>       
>       * I am not sure about the style because I haven't tested this change, but I am sure `muted` is not needed because `muted`  is for `a` only.

Agreed. I just opted to conserve classes. Is now removed.

> 
>     * The `class="dropdown icon"`  classes on SVG elements in dropdowns  is done by purpose to fine-tune the UI alignment, ~there are CSS styles for them, I do not think the `dropdown` could be removed (I haven't tested, just a guess)~
>       
>       * Checked, for other SVG icons, the `svg dropdown icon` is done by purpose.
>       * For this one, it's better to remove the `dropdown` class. This icon already has its own `gt-mt-...` styles.
> 
>     * ~`attachDropdownAria($('.ui.dropdown'));` doens't seem right. We should skip `custom`  dropdown elements, these `custom` elements have their own JS to work.~
>       
>       * Hmm ... old code also did so, maybe I didn't realize this problem at that time.
> 
>     * The new JS code: I will take a look carefully later. A quick thought in my mind: maybe we need two functions, one for `menu` , one for `combobox` (they could share  some  code  by common functions), to make them independent, then any change on one side won't affect another one, and it's easier to maintain and test in the future.

That's what I said above. If you implement combobox based in `input` presence IMO you will produce more technical debt.

>       
>       * I made some new changes in [Fix accessibility behavior for topic management in the repository home, make some dropdown work as combobox #23346](https://github.com/go-gitea/gitea/pull/23346)
>       * According to [w3 demo](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-none/), `combobox` is enough, no need to use `searchbox`

Perfect, but you should need to remove the label inside because is out of that pattern as I explained above with language selector.

>       * Most code could be shared between `combobox` and `menu`

It depends. Several dropdowns have the `input` inside the menu container, which is out of accessible standard.

>       * After some quick tests, I think these changes work well for many cases.
>       * If `#23346` is good enough, could we continue to work on it?
> 
>     * Update: one more thing, the `item` shouldn't be clicked without any check. See the comment in:
>       
>       * [Fix incorrect display for comment context menu  #23343](https://github.com/go-gitea/gitea/pull/23343)

This reversion was already integrated today.

wxiaoguang commented:
Thank you for the details. I can understand more about your thoughts.

To be clear, when I mention `dropdown`, it always means Fomantic UI Dropdown. If a component could have a clear role, I will call it `menu` or `combobox`.

> menu role doesn't have any element except the menuitem set. Language selector has the current value as a text which appears to be a readonly input. This is why I implemented as a combobox.

I really prefer to make `profile` `language` and `create repo` dropdowns as `menu`. IMO  we should choose their roles by their functions and designed behaviors, but not by  what they have or what they did.

The  behaviors of  these dropdowns:

* It has a button, click the button, the menu shows, the menu contains menu-items, if the user clicks a menu-item, an action is executed and the page will be reloaded (the key point). It doesn't provide any input for forms.
* I do not think the `current value` affects that behavior,  in fact, the  `current value` which is changed by click action is an incorrect behavior and should be fixed.
    * Think about a more complex situation, take the  language dropdown as an example, it  used  AJAX to change  language. 
        * If the AJAX  request fails but the `current value` is still changed,  it makes the UI inconsistent, that's a real UI bug.
        * If we make it a menu, the current value never changes. If the AJAX request fails, the text is still the  same as current language. If the  AJAX request  succeeds, the page reloads and it  still shows the correct current language.
* About `auto-submit`: if you are in a form, you need to click a button to submit. But I don't think you need to confirm the language in a language menu, the `profile` and `create repo` are also menus, I do not see why a user need to use them as combobox and confirm before submit.

> That's what I said above. If you implement combobox based in input presence IMO you will produce more technical debt.

It won't cause any technical debt, instead, a simple patch helps much.

In my plan, the `aria.js` is only used for simple dropdowns. Complex dropdowns should have their own code. See below.

> The worst case is the tag/branch selector

That's that case in my proposal "Popup (the branch/tag panel, the review box)", in fact, we should totally rewrite them, instead of patching them. These components couldn't be fixed by a simple patch, I never want to use `aria.js` to fix these complex elements. 

And I have removed the `$.dropdown()` for review box, not that complex, we just need some more time to fix them (ps: I haven't fine tune a11y for it due to insufficient time). 

----

The whole picture  in my mind about how to improve the a11y for dropdown and avoid technical debt:

* Making `aria.js` too complex won't help too much, it won't be perfect solution for all cases.
* Keep `aria.js` as simple as possible, make it only handle common cases:
    * Simple menu (no input)
    * Simple combobox (with input)
    * Improve some  comboboxes if it could, but do not introduce too much specialized code to avoid technical debt
    * Rewrite popup dropdowns (I have remove dropdown from review box)
    * Improve complex dropdowns (like Issue Labels) one by one, because they are so special

wxiaoguang commented:
FYI, to fixe the "branch selector dropdown", I think we need to to this first:  Refactor branch/tag selector dropdown (first step) #23394

wxiaoguang commented:
About the "language menu" / "reaction menu" problem, according to my latest debug, I think it's mainly related the event sequence for different browsers and different screen readers. Fomantic UI has made a mess, I tried to made a new patch in https://github.com/go-gitea/gitea/pull/23346/commits/7a3d1a018af4e8fc141efd1188d3c4b5d54dfff1 .

It's really difficult to do a full test for various cases, I have done my best to test it with my desktop/mobile with/wihout screen readers.

Feel free to take look.

wxiaoguang commented:
I see your question in Discord.

1. The "menu" vs "combobox" concept: we indeed have divergence. I believe some incorrect "menu" behaviors are caused by bugs, so they shouldn't be "combobox".
2. I am not maintainer at the moment, although you have the "Allow Maintainer Edit" checked, I can't help to edit your PR
    * If you agree, I can apply to be a maintainer again to help to edit on this PR directly, resolve all conflicts caused by me

wxiaoguang commented:
More proofs about "why the language dropdown should be a menu".

Take a look at W3 standard:

* Combobox:
  * https://www.w3.org/WAI/ARIA/apg/patterns/combobox/
  * A combobox is an **input widget** with an associated popup that enables users to select a value for the combobox from
    a collection of possible values. In some implementations, the popup presents allowed values, while in other implementations,
    the popup presents suggested values, and users may either select one of the suggestions or type a value.
* Menu:
  * https://www.w3.org/WAI/ARIA/apg/patterns/menubar/
  * A menu is a widget that offers a list of choices to the user, such as a set of **actions or functions**.


And take a look at Apple's page:

![image](https://user-images.githubusercontent.com/2114189/224600326-d3dfa62c-4cd3-4427-bd1a-6d6587811081.png)

wxiaoguang commented:
Stale for long time, I think I have been open and clear for this problem.
